### ISSUE TITLE
Exception in Netif.writev() during ICMP echo request handling

### ISSUE THREAD

**Issue by mato** [mirage/mirage-net-solo5#2]
Solo5 and UNIX unikernels (eg. `stackv4` from mirage-skeleton) built against current mirage master run into the following exception when `ping`ed:

```
$ ping 10.0.0.2
PING 10.0.0.2 (10.0.0.2) 56(84) bytes of data.
^C
--- 10.0.0.2 ping statistics ---
2 packets transmitted, 0 received, 100% packet loss, time 999ms

Netif: [listen] error while handling (Invalid_argument
  "Cstruct.blit src=[42,56](4096) dst=[0,4096](4096) dst-off=4096 len=56"), continuing. bt: Raised at file "pervasives.ml", line 33, characters 25-45
Called from file "lib/netif.ml", line 176, characters 8-38
Called from file "list.ml", line 77, characters 12-15
Called from file "lib/netif.ml", line 174, characters 4-139
Called from file "src/core/lwt.ml", line 653, characters 66-69

Netif: [listen] error while handling (Invalid_argument
  "Cstruct.blit src=[42,56](4096) dst=[0,4096](4096) dst-off=4096 len=56"), continuing. bt: Raised at file "pervasives.ml", line 33, characters 25-45
Called from file "lib/netif.ml", line 176, characters 8-38
Called from file "list.ml", line 77, characters 12-15
Called from file "lib/netif.ml", line 174, characters 4-139
Called from file "src/core/lwt.ml", line 686, characters 20-24
```

Note that this issue does not appear to be solo5-specific, as I can reproduce the exact same exception (bar differences in line numbers) with a `unix` target. However, it may be specific to the Solo5/UNIX Netif layer as `mirage-net-solo5` is derived from `mirage-net-unix`.

What is interesting is that the unikernel handles TCP and ARP fine, only ICMP appears to be affected.

**Comment by mato** [comment]
@hannesm @yomimono Any ideas on this one? Has something changed in the layers `netif.ml` plugs into?

**Comment by hannesm** [comment]
I think there were ICMP related changes in mirage-tcpip, but I've not much clue about details.  @yomimono likely knows details (not sure until when she's in transit).

**Comment by yomimono** [comment]
@hannesm is correct! I'll have a closer look at this later today.

**Comment by yomimono** [comment]
`writev` in mirage-net-unix looks to me to have a bug when the total length of stuff to be written is > the size of 1 page.  A bug in the ICMP write code is triggering this condition. I'm writing a test for mirage-net-unix to confirm that this is indeed broken; I'll then write a fix for ICMP and verify it.

**Comment by yomimono** [comment]
PR for `mirage-net-unix`'s poor handling of this case is at https://github.com/mirage/mirage-net-unix/pulls .  I've removed the custom and broken code path for constructing ICMP echo replies in `tcpip` as well in http://github.com/mirage/mirage-tcpip/pull/231 .

**Comment by mato** [comment]
Fixed in #3.
